persistence/sql/persister_recovery.go

```diff
@@ -127,3 +128,67 @@ func (p *Persister) DeleteExpiredRecoveryFlows(ctx context.Context, expiresAt ti
 	}
 	return nil
 }
+
+func (p *Persister) CreateRecoveryCode(ctx context.Context, recoveryCode *code.RecoveryCode) error {
+	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.CreateRecoveryCode")
+	defer span.End()
+
+	code := recoveryCode.Code
+	recoveryCode.Code = p.hmacValue(ctx, code)
+	recoveryCode.NID = p.NetworkID(ctx)
+
+	// This should not create the request eagerly because otherwise we might accidentally create an address that isn't
+	// supposed to be in the database.
+	if err := p.GetConnection(ctx).Create(recoveryCode); err != nil {
+		return err
+	}
+
+	recoveryCode.Code = code
+	return nil
+}
+
+func (p *Persister) UseRecoveryCode(ctx context.Context, fID uuid.UUID, codeVal string) (*code.RecoveryCode, error) {
+	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.UseRecoveryCode")
+	defer span.End()
+
+	var recoveryCode code.RecoveryCode
+
+	nid := p.NetworkID(ctx)
+	if err := sqlcon.HandleError(p.Transaction(ctx, func(ctx context.Context, tx *pop.Connection) (err error) {
+		for _, secret := range p.r.Config().SecretsSession(ctx) {
+			if err = tx.Where("code = ? AND nid = ? AND NOT used AND selfservice_recovery_flow_id = ?", p.hmacValueWithSecret(ctx, codeVal, secret), nid, fID).First(&recoveryCode); err != nil {
```

Comment:
> Is this using a different way than during creation to compute the hmac?

Yes. `p.hmacValue(ctx, code)` uses the **current** secret and here we need to use all known secrets.

> I propose to instead do a single query

Good idea! Small correction: `nid = ? AND selfservice_recovery_flow_id = ?` can return multiple rows, because you can create multiple codes from one recovery flow. So we'd need to iterate over the rows and do the checks on each one of them. Most of the time, it will just be one code, though. With this change, we can be more specific with the errors as well.

Reply:
OK nice, then let's go with that single query!

Reply:
Done.